src/ngraph/runtime/gpu/cuda_emitter.cpp

```diff
@@ -677,6 +678,95 @@ size_t runtime::gpu::CUDAEmitter::build_elementwise_n_to_1(const GPURuntimeConte
+                                                             const Shape& input_shape,
+                                                             const Shape& output_shape,
+                                                             const Shape& reduce_window_shape,
+                                                             const Strides& reduce_window_strides,
+                                                             const char* op,
+                                                             const char* kernel)
+{
+    // kernel_name is used to check if the cuda kernel has been previously compiled
+    size_t rank = input_shape.size();
+    std::stringstream kernel_name;
+    kernel_name << "reduce_window"
+                << "_" << op << "_" << join(dtypes, "_") << rank;
+
+    // hash is used to check if the emitted primitive already exists
+    std::stringstream ss;
+    ss << kernel_name.str() << "_s" << join(output_shape, "_");
+    auto hash = ss.str();
+
+    // if the primitive exists, we are done
+    size_t primitive_index = m_primitive_emitter->lookup(hash);
+    if (primitive_index != std::numeric_limits<size_t>::max())
+    {
+        return primitive_index;
+    }
+
+    // check if the kernel has already been compiled. if so, create
+    // a launch primitive for it based on the input tensor shape
+    // but do not recompile the kernel. otherwise, do it all:
+    // recompile the kernel and then create the primitive
+    auto compiled_kernel = ctx->compiled_kernel_pool->get(kernel_name.str());
+    if (compiled_kernel == nullptr)
+    {
+        codegen::CodeWriter writer;
+        CudaKernelBuilder::add_pod_typedefs(writer);
+        if (kernel)
+        {
+            CudaKernelBuilder::get_device_helper(writer, op, kernel, dtypes);
+        }
+
+        CudaKernelBuilder::get_reduce_window_op(writer, kernel_name.str(), op, dtypes, rank);
+
+        compiled_kernel = ctx->compiled_kernel_pool->set(kernel_name.str(), writer.get_code());
+    }
+    size_t nthreads = shape_size(output_shape);
+    auto input_strides = row_major_strides(input_shape);
+
+    std::shared_ptr<void*> input_strides_ptr = std::make_shared<void*>(
+        runtime::gpu::create_gpu_buffer(sizeof(size_t) * rank, input_strides.data()));
+    std::shared_ptr<void*> output_shape_ptr = std::make_shared<void*>(
+        runtime::gpu::create_gpu_buffer(sizeof(size_t) * rank, output_shape.data()));
+    std::shared_ptr<void*> reduce_window_shape_ptr = std::make_shared<void*>(
+        runtime::gpu::create_gpu_buffer(sizeof(size_t) * rank, reduce_window_shape.data()));
+    std::shared_ptr<void*> reduce_window_stride_ptr = std::make_shared<void*>(
+        runtime::gpu::create_gpu_buffer(sizeof(size_t) * rank, reduce_window_strides.data()));
+
+    // create the launch primitive
+    std::unique_ptr<gpu::primitive> f(
+        new gpu::primitive{[=](void** inputs, void** outputs) mutable {
+            std::vector<void*> args_list;
```

Comment:
nit: Why use std::vector::push_back() here when the number of elements is fixed and known at compile time? std::vector<void*> args_list { ... } would be better so that you avoid the copying when push_back overflows the internal memory storage of the vector.